### PR TITLE
Serializing samplers with scenario constants configuration

### PR DESCRIPTION
**Pull request by sleal-unity**
# Peer Review Information:

**NOTICE**: Merge the [global-seed PR](https://github.com/Unity-Technologies/com.unity.perception/pull/142) first before merging this PR.

This PR extends the serialization functionality of the Scenario's constants configuration to serialize samplers in randomizers along with the original list of scenario constants.

Other changes include improving UI and console feedback for configuration file serialization and deserialization. The UI will now ping the newly created configuration file in the project window when it is created. When deserializing a configuration file, a notification will appear in the console informing users that all constants and randomizer changes are undo-able.

Configuration files will no longer be deserialized automatically in playmode (users must manually click the deserialize button before entering playmode). However, configuration files will be deserialized automatically if present for built players.

The new configuration file format is also compatible with Unity Simulation runs.

Below are two serialized json configurations generated from the perception tutorial before and after including this PR:

#### Before:
```
{
  "framesPerIteration": 1,
  "totalIterations": 100,
  "instanceCount": 1,
  "instanceIndex": 0,
  "randomSeed": 123456789
}
```

#### After:

```
{
  "constants": {
    "framesPerIteration": 1,
    "totalIterations": 100,
    "instanceCount": 1,
    "instanceIndex": 0,
    "randomSeed": 123456789
  },
  "randomizers": {
    "HueOffsetRandomizer": {
      "hueOffset": {
        "value": {
          "range": {
            "minimum": -180.0,
            "maximum": 180.0
          }
        }
      }
    },
    "RotationRandomizer": {
      "rotation": {
        "x": {
          "range": {
            "minimum": 0.0,
            "maximum": 360.0
          }
        },
        "y": {
          "range": {
            "minimum": 0.0,
            "maximum": 360.0
          }
        },
        "z": {
          "range": {
            "minimum": 0.0,
            "maximum": 360.0
          }
        }
      }
    }
  }
}
```

## Editor / Package versioning:
**Editor Version Target**: 2019.4

## Dev Testing:
**Tests Added**: 
Added additional serialization tests to confirm that randomizers and constants serialize and deserialize properly.

## Checklist
- [X] - Updated docs
- [X] - Updated changelog

### CODE REVIEW

File: com.unity.perception/Editor/Randomization/Editors/RunInUnitySimulationWindow.cs [hunk 1]
```diff
@@ -186,24 +188,30 @@ void CreateLinuxBuildAndZip()
         List<AppParam> GenerateAppParamIds(CancellationToken token)
         {
             var appParamIds = new List<AppParam>();
+            var scenario = (ScenarioBase)m_ScenarioField.value;
+            var configuration = JObject.Parse(scenario.SerializeToJson());
+            var constants = configuration["constants"];
```
Comment (mkamalza) [inline com.unity.perception/Editor/Randomization/Editors/RunInUnitySimulationWindow.cs:193]: What is the use of this constants variable defined here?

**Review by mkamalza — CHANGES_REQUESTED** [review]
Great job. Just a couple of minor things.

File: com.unity.perception/Editor/Randomization/Editors/RunInUnitySimulationWindow.cs (hunk 1 above)
Comment (sleal-unity) [reply com.unity.perception/Editor/Randomization/Editors/RunInUnitySimulationWindow.cs:193]: The constants variable here is a reference to the constant's section of the scenario json configuration. The loop below assigns the appropriate instance index to each created app-param by leveraging this reference.

Comment (sleal-unity) [reply com.unity.perception/Editor/Randomization/Editors/RunInUnitySimulationWindow.cs:193]: This constants section is consistent with the naming of the constants field defined for every scenario that inherits from the Scenario<T> abstract class.

**Review by mkamalza — APPROVED** [review]

**Pull request merged**